component/featuregate.go

```diff
@@ -0,0 +1,40 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package component // import "go.opentelemetry.io/collector/component"
+
+import (
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/featuregate"
+)
+
+const useLocalHostAsDefaultHostID = "component.UseLocalHostAsDefaultHost"
+
+// UseLocalHostAsDefaultHostfeatureGate is the feature gate that controls whether
+// server-like receivers and extensions such as the OTLP receiver use localhost as the default host for their endpoints.
+var UseLocalHostAsDefaultHostfeatureGate = featuregate.GlobalRegistry().MustRegister(
+	useLocalHostAsDefaultHostID,
+	featuregate.StageAlpha,
+	featuregate.WithRegisterDescription("controls whether server-like receivers and extensions such as the OTLP receiver use localhost as the default host for their endpoints"))
+
+// EndpointForPort gets the endpoint for a given port using localhost or 0.0.0.0 depending on the feature gate.
+func EndpointForPort(port int) string {
+	host := "0.0.0.0"
+	if UseLocalHostAsDefaultHostfeatureGate.IsEnabled() {
```

Comment:
Instead of setting this to `localhost`, would it make sense to allow setting whatever value from the feature flag?

Reply:
The feature gate is meant to ease migration to the new defaults, but I wouldn't want people relying on it to set the endpoint. For existing users, if they want a different endpoint then they have already modified the YAML, so I don't think it affects them. For new users, we want them to modify the YAML instead of using the gate. So I don't think we should add this capability.